library/src/G4CMPPhononBoundaryProcess.cc

```diff
@@ -192,128 +227,249 @@ DoReflection(const G4Track& aTrack, const G4Step& aStep,
     G4Track* sec1 = particleChange.GetSecondary(0);
     G4Track* sec2 = particleChange.GetSecondary(1);
 
-    G4ThreeVector vec1 = GetLambertianVector(surfNorm, mode);
-    G4ThreeVector vec2 = GetLambertianVector(surfNorm, mode);
+    G4ThreeVector vec1 = G4CMP::GetLambertianVector(theLattice, surfNorm, mode,
+                                                    surfacePoint);
+    G4ThreeVector vec2 = G4CMP::GetLambertianVector(theLattice, surfNorm, mode,
+                                                    surfacePoint);
 
     sec1->SetMomentumDirection(vec1);
     sec2->SetMomentumDirection(vec2);
 
     return;
   } else if (random < downconversionProb + specProb) {
-    reflectedKDir = GetReflectedVector(waveVector, surfNorm, mode);
+    reflectedKDir = GetSpecularVector(waveVector, surfNorm, mode, surfacePoint); // Modify surfacePoint & surfNorm in place
     refltype = "specular";
   } else {
-    reflectedKDir = GetLambertianVector(surfNorm, mode);
+    reflectedKDir = G4CMP::GetLambertianVector(theLattice, surfNorm, mode,
+                                               surfacePoint);
     refltype = "diffuse";
   }
 
-  G4ThreeVector vdir = theLattice->MapKtoVDir(mode, reflectedKDir);
-  G4double v = theLattice->MapKtoV(mode, reflectedKDir);
+  // Update trackInfo wavevector and particleChange's group velocity and momentum direction
+  // reflectedKDir is in global coordinates here - no conversion needed
+  FillParticleChange(particleChange, aTrack, reflectedKDir);
+  const G4ThreeVector vdir = *particleChange.GetMomentumDirection();
+
+  // If displacement occured: update the particle change and navigator for volume assignment
+  if (refltype == "specular" && *particleChange.GetPosition() != surfacePoint) {
+    FillParticleChange(phParticleChange, aStep, surfacePoint);
+    UpdateNavigatorVolume(aStep, surfacePoint, vdir);
+  }
 
   if (verboseLevel>2) {
-    G4cout << "\n New wavevector direction " << reflectedKDir
-	   << "\n New momentum direction   " << vdir << G4endl;
+    G4cout << "New surface position " << *particleChange.GetPosition()/mm 
+	   << " mm\n New wavevector direction " << reflectedKDir
+	   << "\n New momentum direction " << vdir << G4endl;
   }
 
   // If reflection failed, report problem and kill the track
-  if (!G4CMP::PhononVelocityIsInward(theLattice,mode,reflectedKDir,surfNorm)) {
+  if (!G4CMP::PhononVelocityIsInward(theLattice,mode,reflectedKDir,surfNorm, surfacePoint)) {
     G4Exception((GetProcessName()+"::DoReflection").c_str(), "Boundary010",
-		JustWarning, ("Phonon "+refltype+" reflection failed").c_str());
-    DoSimpleKill(aTrack, aStep, aParticleChange);
+		JustWarning, ("Phonon "+refltype+" reflection failed"+"\nPhonon mode at time of death: "+G4PhononPolarization::Label(mode)).c_str());
```

Comment:
The "\n" can be in the middle of the string literal, which might be better depending on how a given compiler does grouping.  And maybe break it out to a temporary variable, to allow line breaks:
```
G4String msg = "Phonon "+refltype+" reflection failed";
msg += "\nPhonon mode at time of death: " + G4PhononPolarization::Label(mode);
```
Then you can have `msg.c_str()` in the `G4Exception()` call.